### PR TITLE
Prevent Rust feature bleed

### PR DESCRIPTION
**Pull request by ggwpez**
See https://github.com/paritytech/substrate/pull/12341 for a description of the new script.  

Currently working in Substrate: [success case](https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1925727) and [error case](https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1925764).

### CODE REVIEW

**Review by mordamax — APPROVED** [review]

File: rust-features.sh [hunk 1]
```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
```
Comment (bkchr) [inline rust-features.sh:1]:
```suggestion
#!//usr/bin/env bash
```

There is no convention that `/bin/bash` exists.

**Comment by bkchr** [comment]
@mordamax you are fine with maintaining this? My eyes bleed again a little bit after seeing this bash script.

**Comment by bkchr** [comment]
BTW, there also exist this pr: https://github.com/paritytech/diener/pull/19

**Comment by mordamax** [comment]
> @mordamax you are fine with maintaining this? My eyes bleed again a little bit after seeing this bash script.

I guess it's @paritytech-ci who maintain it in this case.
This repo is mainly for reusable CI scripts, to avoid dupe

but my opinion is that it's okay as long as it's relevantly simple, <=~100 lines

the existing stuff like check-dependent should be killed 100%
tryruntime and bench most probably will be moved & rewritten to command bot

**Comment by ggwpez** [comment]
> BTW, there also exist this pr: [paritytech/diener#19](https://github.com/paritytech/diener/pull/19)

Interesting... I dont think parsing `Cargo.toml` files works for external dependencies from `crates.io`, or? (In case we ever release Substrate there :rofl:) But in any case, that check is additive to this one.

I am not happy with this as long-term solution, but for now it should prevent obvious blunders which can take a while to find otherwise. Recently I saw https://github.com/guppy-rs/guppy which is used by Diem and think could be a fit.

**Comment by bkchr** [comment]
> I dont think parsing `Cargo.toml` files works for external dependencies from `crates.io`, or

You download the deps to the local cargo registry. So, this should work.



> Recently I saw https://github.com/guppy-rs/guppy which is used by Diem and think could be a fit.

I think I read about this somewhere before, but if that could be a fit, it would be nice.

**Comment by ggwpez** [comment]
Okay then lets use this to get rid of some obvious errors and then make a long-term strategy?  
Like set out a scope of rules that we want to be checked and search for existing tools if they can do it or self implement.

File: rust-features.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
+#!//usr/bin/env bash
```
Comment (mutantcornholio) [inline rust-features.sh:1]: Is this double leading accidental?

Comment (ggwpez) [inline rust-features.sh:1]:
```suggestion
#!/usr/bin/env bash
```
Nope but it would still work :smile:

Comment (ggwpez) [reply rust-features.sh]: fixed

File: rust-features.sh [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+for RULE in "${FEATURE_RULES[@]}"; do
+    read -a splits <<< "$RULE"
+	ENABLED=${splits[0]}
+	STAYS_DISABLED=${splits[4]}
+	# Split ENABLED at , and check each one.
+	IFS=',' read -ra ENABLED_SPLIT <<< "$ENABLED"
+	for ENABLED_SPLIT in "${ENABLED_SPLIT[@]}"; do
+		check_does_not_imply "$ENABLED_SPLIT" "$STAYS_DISABLED"
+	done	
+done
```
Comment (mutantcornholio) [inline rust-features.sh:96]: This feels too complicated to be something that processes static data. Maybe we could organise data differently?  
The `"default,std never implies feature try-runtime"` syntax also seems hacky.

File: rust-features.sh [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+	# Find all Cargo.toml files but exclude the root one since we know that it is broken.
+	CARGOS=`find . -name Cargo.toml -not -path ./Cargo.toml`
+	FAILED=0
+	PASSED=0
+	# number of all cargos
+	echo "🔍 Checking individual crates - this takes some time."
+
+	for CARGO in $CARGOS; do
+		RET=0
```
Comment (mutantcornholio) [inline rust-features.sh:63]: I genuinely can't understand what's going on here...  
We have a variable that stores return code of `cargo tree`, and then we're nullifying it in the loop... But how is it used afterwards?

File: rust-features.sh [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+	RET=0
+	# Check if the forbidden feature is enabled anywhere in the workspace.
+	cargo tree --no-default-features --locked --workspace -e features --features "$ENABLED" | grep -q "feature \"$STAYS_DISABLED\"" || RET=$?
+	if [ $RET -ne 0 ]; then
+		echo "✅ $ENABLED does not imply $STAYS_DISABLED in the workspace"
+		return
+	else
+		echo "❌ $ENABLED implies $STAYS_DISABLED in the workspace"
+	fi
```
Comment (mutantcornholio) [inline rust-features.sh:53]: Can't it be simplified like this?

```suggestion
	# Check if the forbidden feature is enabled anywhere in the workspace.
	if cargo tree --no-default-features --locked --workspace -e features --features "$ENABLED" | grep -q "feature \"$STAYS_DISABLED\""; then
		echo "✅ $ENABLED does not imply $STAYS_DISABLED in the workspace"
		return
	else
		echo "❌ $ENABLED implies $STAYS_DISABLED in the workspace"
	fi
```

Also, `-F` is probably implied here in `grep` invocation.

File: rust-features.sh [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+#   5. Print all offending crates and exit with code 1.
+#
+##############################################################################
+
+set -eu
+
+# Check that cargo, grep and egrep are installed.
+command -v cargo >/dev/null 2>&1 || { echo >&2 "cargo is required but not installed. Aborting."; exit 1; }
+command -v grep >/dev/null 2>&1 || { echo >&2 "grep is required but not installed. Aborting."; exit 1; }
+command -v egrep >/dev/null 2>&1 || { echo >&2 "egrep is required but not installed. Aborting."; exit 1; }
```
Comment (mutantcornholio) [inline rust-features.sh:29]: is `egrep` used here somewhere?

File: rust-features.sh [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+CARGO_ROOT=$1
+cd "$CARGO_ROOT"
+
+# NOTE: The features should be separated by a single `,`.
+declare -a FEATURE_RULES=(
+	"default,std never implies feature runtime-benchmarks"
+	"default,std never implies feature try-runtime"
```
Comment (mutantcornholio) [inline rust-features.sh:37]: After spending time on reviewing this, I'd say this format ought to be changed.  
It's not self-descriptive, the fact that it expects "feature<space><feature_name>" at the end doesn't feel natural, and a lot of the code here is dedicated specifically to parsing of the format.

**Review by mutantcornholio — CHANGES_REQUESTED** [review]

File: rust-features.sh (hunk 3 above)
Comment (ggwpez) [reply rust-features.sh:96]: Yea I can put it in some arrays instead. But then it needs more docs on how to use it. Like this it reads as natural sentence.

File: rust-features.sh (hunk 7 above)
Comment (ggwpez) [reply rust-features.sh:37]: Okay I will simplify it.

File: rust-features.sh (hunk 3 above)
Comment (mutantcornholio) [reply rust-features.sh:96]: I believe that one comment line describing the format, would save a lot more time to the reader of the script than this idea of natural sentence, which actually not natural at all, but a strict format instead.

Comment (ggwpez) [reply rust-features.sh:96]: I removed the data structure and parsing, should be much easier now :see_no_evil:

File: rust-features.sh (hunk 6 above)
Comment (ggwpez) [reply rust-features.sh]: not anymore

File: rust-features.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+	for CARGO in $CARGOS; do
+		OUTPUT=$(cargo tree --no-default-features --locked --offline -e features --features $ENABLED --manifest-path $CARGO 2>&1 || true)
+		IS_NOT_SUPPORTED=$(echo $OUTPUT | grep -qF "not supported for packages in this workspace" || echo $?)
+
+		if [ $IS_NOT_SUPPORTED -eq 0 ]; then
+			# This case just means that the pallet does not support the
+			# requested feature which is fine.
+			PASSED=$((PASSED+1))
+		elif echo "$OUTPUT" | grep -qF "feature \"$STAYS_DISABLED\""; then
```
Comment (mutantcornholio) [inline rust-features.sh:71]: I see three different approaches to checking return code here.  
This one, inside `if`, then `RET=0; command || RET=$?`, and `IS_NOT_SUPPORTED=$(command || echo $?)`.  
Maybe put all of them into `if`'s?

Comment (ggwpez) [reply rust-features.sh:71]: I kind of wanted to keep the lines short, but I guess I can break it into multiple like this as well :+1:

File: rust-features.sh [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+			PASSED=$((PASSED+1))
+		fi
+	done
+
+	echo "Checked $NUM_CRATES crates in total of which $FAILED failed and $PASSED passed."
+	echo "Exiting with code 1"
+	exit 1
+}
+
+main "$@"; exit
```
Comment (mutantcornholio) [inline rust-features.sh]: what's the purpose of `exit` here?

Comment (ggwpez) [reply rust-features.sh]: I just removed it, should be not needed.

**Review by mutantcornholio — APPROVED** [review]
anyway, much more readable and maintainable now, thank you!

**Pull request merged**